File: 2021/2021W37/preppindataW37.R
library(readxl)
library(dplyr)
library(lubridate)

contract <- read_excel("F:/Data/2021 Week 37 Input.xlsx", sheet = "Contract Details")
                
final <- contract[rep(seq_len(nrow(contract)), contract[['Contract Length (months)']]),] %>%
  group_by(`Name`) %>%
  mutate('rownum' = row_number(`Name`),
         'Payment Date' = `Start Date` %m+% months(`rownum`-1),
         'Cumulative Monthly Cost' = `Monthly Cost`*`rownum`) %>%
  select(c('Name', 'Payment Date', 'Monthly Cost', 'Cumulative Monthly Cost'))

View(final)

File: 2021/2021W37/PreppindataW37.py
import pandas as pd
import numpy as np

xlsx = pd.ExcelFile(r"F:\Data\2021 Week 37 Input.xlsx")

contract = pd.read_excel(xlsx,'Contract Details')
final = contract.iloc[np.repeat(np.arange(len(contract)), list(contract['Contract Length (months)']))].copy()
final['rownum'] = final.groupby(['Name'])['Start Date'].rank(method='first')
final['Payment Date'] = final.apply(lambda x: (x['Start Date']+pd.DateOffset(months=x['rownum']-1)).date(), axis=1) 
final['Cumulative Monthly Cost'] = final['Monthly Cost']*final['rownum'] 
final = final[['Name', 'Payment Date', 'Monthly Cost', 'Cumulative Monthly Cost']]
